beacon-chain/casper/validator.go

```diff
@@ -184,3 +184,95 @@ func TotalActiveValidatorDepositInEth(validators []*pb.ValidatorRecord) uint64 {
 
 	return depositInEth
 }
+
+// AddValidator runs for every validator that is inducted as part of a log created on the PoW chain.
+func AddValidator(
+	validators []*pb.ValidatorRecord,
+	pubKey []byte,
+	withdrawalShard uint64,
+	withdrawalAddr []byte,
+	randaoCommitment []byte) []*pb.ValidatorRecord {
+
+	// TODO(#633): Use BLS to verify signature proof of possession and pubkey and hash of pubkey.
+
+	newValidatorRecord := &pb.ValidatorRecord{
+		Pubkey:            pubKey,
+		WithdrawalShard:   withdrawalShard,
+		WithdrawalAddress: withdrawalAddr,
+		RandaoCommitment:  randaoCommitment,
+		Balance:           uint64(params.GetConfig().DepositSize * params.GetConfig().Gwei),
+		Status:            uint64(params.PendingActivation),
+		ExitSlot:          0,
+	}
+
+	index := minEmptyValidator(validators)
+	if index > 0 {
+		validators[index] = newValidatorRecord
+		return validators
+	}
+
+	validators = append(validators, newValidatorRecord)
+	return validators
+}
+
+// ChangeValidators updates the validator set during state transition.
+func ChangeValidators(currentSlot uint64, totalPenalties uint64, validators []*pb.ValidatorRecord) []*pb.ValidatorRecord {
+	maxAllowableChange := uint64(2 * params.GetConfig().DepositSize * params.GetConfig().Gwei)
+
+	totalBalance := TotalActiveValidatorDeposit(validators)
+
+	// Determine the max total wei that can deposit and withdraw.
+	if totalBalance > maxAllowableChange {
+		maxAllowableChange = totalBalance
+	}
+
+	var totalChanged uint64
+	for i := 0; i < len(validators); i++ {
+		if validators[i].Status == uint64(params.PendingActivation) {
+			validators[i].Status = uint64(params.Active)
+			totalChanged += uint64(params.GetConfig().DepositSize * params.GetConfig().Gwei)
+			// TODO(#614): Add validator set change.
+		}
+		if validators[i].Status == uint64(params.PendingExit) {
+			validators[i].Status = uint64(params.PendingWithdraw)
+			validators[i].ExitSlot = currentSlot
+			totalChanged += validators[i].Balance
+			// TODO(#614): Add validator set change.
+		}
+		if totalChanged > maxAllowableChange {
+			break
+		}
+	}
+
+	// Calculate withdraw validators that have been logged out long enough,
+	// apply their penalties if they were slashed.
+
+	for i := 0; i < len(validators); i++ {
+		if (validators[i].Status == uint64(params.PendingWithdraw) ||
```

Comment:
This is a massive conditional - consider splitting into variables that have clear names such as isPendingWithdraw or isPenalized, etc.